src/ray/util/shared_lru.h

```diff
@@ -0,0 +1,135 @@
+// SharedLruCache is a LRU cache, with all entries shared, which means a single entry
+// could be accessed by multiple getters. When `Get`, a copy of the value is returned, so
+// for heavy-loaded types it's suggested to wrap with `std::shared_ptr<>`.
+//
+// Example usage:
+// SharedLruCache<std::string, std::string> cache{cap};
+// // Put a key-value pair into cache.
+// cache.Put("key", "val");
+//
+// // Get a key-value pair from cache.
+// auto val = ache.Get("key");
+// // Check and consume `val`.
+//
+// TODO(hjiang):
+// 1. Add template arguments for key hash and key equal, to pass into absl::flat_hash_map.
+// 2. Provide a key hash wrapper to save a copy.
+// 3. flat hash map supports heterogeneous lookup, expose `KeyLike` templated interface.
+
+#pragma once
+
+#include <cstdint>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <string>
+#include <utility>
+
+#include "absl/container/flat_hash_map.h"
+
+namespace ray::utils::container {
+
+template <typename Key, typename Val>
+class SharedLruCache final {
+ public:
+  using key_type = Key;
+  using mapped_type = Val;
+
+  // A `max_entries` of 0 means that there is no limit on the number of entries
+  // in the cache.
+  explicit SharedLruCache(size_t max_entries) : max_entries_(max_entries) {}
+
+  ~SharedLruCache() = default;
+
+  // Insert `value` with key `key`. This will replace any previous entry with
+  // the same key.
+  void Put(Key key, Val value) {
+    std::lock_guard lck(mu_);
+    lru_list_.emplace_front(key);
+    Entry new_entry{std::move(value), lru_list_.begin()};
+    cache_[key] = std::move(new_entry);
```

Comment:
Emmm could you please tell me what's the benefit?
- `key` is a constant reference, so move has no effect
- `new_entry` is moved as well, so no move wasted operation
- I try to avoid squeeze all expression in one statement for better readability

Reply:
Actually your implementation is buggy, since `emplace` does overwrite existing entries.
Reference: https://en.cppreference.com/w/cpp/container/unordered_map/emplace

Reply:
oh didn't realize that, but ya readability vs micro-opt of extra move, don't mind either way